src/Workspaces/Remote/Core/BrokeredServiceConnection.cs

```diff
@@ -143,26 +146,76 @@ public override async ValueTask<Optional<TResult>> TryInvokeAsync<TResult>(Solut
-            // See https://github.com/microsoft/vs-streamjsonrpc/blob/master/doc/oob_streams.md
-            var (clientStream, serverStream) = FullDuplexStream.CreatePair();
+            var pipe = new Pipe();
 
-            // Create new tasks that both start executing, rather than invoking the delegates directly.
-            // If the reader started synchronously reading before the writer task started it would be blocking, and vice versa
-            // if the writer synchronously filled the buffer before the reader task started it would also be blocking.
-            var writerTask = Task.Run(async () => await invocation(service, serverStream, cancellationToken).ConfigureAwait(false), cancellationToken);
-            var readerTask = Task.Run(async () => await reader(clientStream, cancellationToken).ConfigureAwait(false), cancellationToken);
-            await Task.WhenAll(writerTask, readerTask).ConfigureAwait(false);
+            // Create new tasks that both start executing, rather than invoking the delegates directly
+            // to make sure both invocation and reader start executing and transfering data.
+
+            var writerTask = Task.Run(async () =>
+            {
+                try
+                {
+                    await invocation(service, pipe.Writer, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    // Ensure that the writer is complete if an exception is thrown
+                    // before the writer is passed to the RPC proxy. Once it's passed to the proxy 
+                    // the proxy should complete it as soon as the remote side completes it.
+                    await pipe.Writer.CompleteAsync(e).ConfigureAwait(false);
+
+                    throw;
+                }
+            }, cancellationToken);
+
+            var readerTask = Task.Run(
+                async () =>
+                {
+                    Exception? exception = null;
+
+                    try
+                    {
+                        return await reader(pipe.Reader, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (Exception e) when ((exception = e) == null)
+                    {
+                        throw ExceptionUtilities.Unreachable;
+                    }
+                    finally
+                    {
+                        await pipe.Reader.CompleteAsync(exception).ConfigureAwait(false);
+                    }
+                }, cancellationToken);
+
+            try
+            {
+                await Task.WhenAll(writerTask, readerTask).ConfigureAwait(false);
+            }
+            catch (AggregateException aggregateException) when (cancellationToken.IsCancellationRequested)
+            {
+                aggregateException.Handle(e => e is OperationCanceledException);
+                cancellationToken.ThrowIfCancellationRequested();
+            }
```

Comment:
📝 This is not reachable. When you `await` a task, the `GetResult` method of the awaiter will unwrap the first inner exception of an aggregate exception which in this case will be whatever exception `writerTask` threw (if any), otherwise whatever `readerTask` threw.